keps/sig-network/20190423-service-lb-finalizer.md

```diff
@@ -0,0 +1,122 @@
+
+Note that the removal of this finalizer might also happen when service type changes from `LoadBalancer` to another. This however doesn't change the implication that the resources cleanup must be fulfilled before fully deleting the service.
+
+The lifecyle of a `LoadBalancer` type serivce with finalizer would look like:
+- Creation
+  1. User creates a service.
+  2. Service controller observes the creation and attaches finalizer to the service.
+  3. Provision of load balancing resources.
+- Deletion
+  1. User issues a deletion for the service.
+  2. Service resoure deletion is blocked due to the finalizer.
+  3. Service controller observed the deletion timestamp is added.
+  4. Cleanup of load balancing resources.
+  5. Service controller removes finalizer from the service.
+  6. Service resource deleted.
+- Update to another type
+  1. User update service from `type=LoadBalancer` to another.
+  2. Service controller observed the update.
+  3. Cleanup of load balancing resources.
+  4. Service controller removes finalizer from the service.
+
+The expected cluster upgrade/downgrade path for service with finalizer would be:
+- Upgrade from pre-finalizer version
+  - All existing `LoadBalancer` services will be attached a finalzer upon startup of the new version of service controller.
+  - The newly created `LoadBalancer` services will have finalizer attached upon creation.
+- Downgrade from with-finailzer version:
+  - All existing `LoadBalancer` service will have the attached finalizer removed upon the cleanup of load balancing resources.
+  - The newly created `LoadBalancer` services will not have finailzer attached.
+
+To ensures that downgrades can happen safely, the first release will include the "remove finalizer" logic with the "add finalizer" logic behind a gate. Then in a later release we will remove the feature gate and enable both the "remove" and "add" logic by default.
+
+As such, we are proposing Alpha/Beta/GA phase for this enhancement as below:
+- Alpha: The finailizer cleanup logic would be on by default while the finalizer addition logic will be put behind a feature gate that is off by default.
+- Beta: Both the cleanup and addition logic will be default on, but allowing disabling via feature gate.
+- GA: Service LoadBalancers Finalizer Protection will always be on.
+
+### Test Plan
+
+We will implement e2e test cases to ensure:
+- Service finalizer protection works with various service lifecycles on a cluster that enables it.
+
+In addition to above, we should have upgrade/downgrade tests that:
+- Verify the downgrade path and ensures service finalizer removal works.
+- Verify the upgrade path and ensures finalizer protection works with existing LB services. 
+
+### Graduation Criteria
+
+Beta: Allow Alpha ("remove finalzier") to soak for at least one release, then switch the "add finalizer" logic to be on by default.
```

Comment:
Wondering if feature gating is necessary here. Can't think of a scenario where you would opt-out of this except for the first version where it's supported. Even if user does a version downgrade, your cluster would break if you forget to remove the feature gate.

Reply:
That's fair point. I was just following the usual lifecycle of a feature gate --- maybe we won't need that for this? Or we can decide when we actually hit beta.

Reply:
My personal take here is to _not_ do feature gates. I consider this more of a bug fix then an added feature.

Reply:
To be clear, did you mean not doing feature gate for this enhancement at all, or not doing feature gate at beta and after?

I think having a feature gate at least in the initial release might help if anyone would like to turn it on earlier (for the "add finalizer" part).

Reply:
>  think having a feature gate at least in the initial release might help if anyone would like to turn it on earlier (for the "add finalizer" part).

Good point 🤔. If we're okay with cleaning up the feature gate after only 1 release then sgtm.